File: tests/oracle/block_domroot_1204.cc
/*
 * HTTPD-ADDDESCEND-THROW-0001: locked Ghidra 12.0.4 dominator-root oracle.
 *
 * Exercises BlockGraph::structureLoops (block.cc:2194-2215 ->
 * findSpanningTree block.cc:1009-1136) followed by
 * BlockGraph::calcForwardDominator (block.cc:1954-2032) on four synthetic
 * graphs chosen to pin the root-selection semantics:
 *
 *   A. multi-root graph with trailing orphan blocks (in0/out0) appended at
 *      the END of the component vector — the httpd lifter shape. Pins the
 *      rootlist head/tail swap (block.cc:1031-1035): the original head
 *      (first sizeIn()==0 block in list order) must finish the DFS last and
 *      take reverse-post-order slot 0, and the final swap (block.cc:1129)
 *      must restore it to rootlist front. Orphans may never steal RPO[0].
 *   B. single root whose entry has a back-edge into it — pins the
 *      createVirtualRoot branch at block.cc:1978-1984 (root with in-edges
 *      gets an artificial root; entry ends with immed_dom == null).
 *   C. cross-root merge — pins the finger-arithmetic alias: a freshly
 *      constructed FlowBlock has index == 0 (block.cc:61-69), so a finger
 *      walking into the virtual root lands on numnodes-0 = the postorder
 *      slot of list[0]. A merge of two different roots' subtrees therefore
 *      converges to immed_dom == list[0], NOT null.
 *   D. pure cycle with no sizeIn()==0 candidate — pins the rootlist
 *      fallback rootlist = { list[0] } (block.cc:1036-1038).
 *
 * After structureLoops + calcForwardDominator the fixture prints, for each
 * case: the reverse post order (component list order after
 * `list = rpostorder`, block.cc:1135), every block's index field, the final
 * rootlist, and every block's immediate dominator (or NULL). All four cases
 * are reducible, so findIrreducible (block.cc:1147) never forces a rebuild
 * and calcLoop (block.cc:2104) never runs — that machinery is outside this
 * fixture's coverage and registered as such in the metadata.
 *
 * Test-only access: BlockGraph::addBlock / newBlock (block.hh:366-413) are
 * implicitly private (the class body never emits an explicit access label
 * before them), so the common `#define private public` fixture lever cannot
 * reach them; production graphs are assembled by Funcdata's flow-following
 * constructor. This fixture additionally defines `class` as `struct` around
 * the decompile headers — flipping the default member access to public —
 * solely to append plain FlowBlock components. structureLoops,
 * calcForwardDominator, getBlock, getIndex and getImmedDom are public API
 * used unmodified. The header set and define order follow the proven
 * block_index_assign_1204.cc pattern: <bits/stdc++.h> prelude, both defines,
 * architecture.hh first (the canonical umbrella include that establishes
 * the declaration order every decompile .cc relies on), then block.hh and
 * the rest of the umbrella headers.
 */
#include <bits/stdc++.h>

#define private public
#define class struct
#include "architecture.hh"
#include "block.hh"
#include "database.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "marshal.hh"
#include "op.hh"
#include "translate.hh"
#include "type.hh"
#include "typeop.hh"
#include "varnode.hh"
#undef class
#undef private

namespace {

using namespace ghidra;

struct CaseGraph {
  BlockGraph graph;
  std::vector<FlowBlock *> blocks;
  std::vector<std::string> names;
  std::map<FlowBlock *, std::string> label_of;

  FlowBlock *add(const std::string &name)
  {
    // newBlock() (block.cc:1659-1665) already registers the new plain
    // FlowBlock via addBlock internally — calling addBlock again would
    // double-register the component.
    FlowBlock *bl = graph.newBlock();
    blocks.push_back(bl);
    names.push_back(name);
    label_of[bl] = name;
    return bl;
  }

  void edge(FlowBlock *from, FlowBlock *to) { graph.addEdge(from, to); }
};

void run_case(const std::string &title, CaseGraph &cg)
{
  std::cout << title << std::endl;

  std::vector<FlowBlock *> rootlist;
  cg.graph.structureLoops(rootlist);
  cg.graph.calcForwardDominator(rootlist);

  // Reverse post order == component list order after `list = rpostorder`.
  std::cout << "rpo:";
  for(int4 i = 0; i < cg.graph.getSize(); ++i) {
    FlowBlock *bl = cg.graph.getBlock(i);
    std::cout << ' ' << cg.label_of[bl] << '(' << bl->getIndex() << ')';
  }
  std::cout << std::endl;

  std::cout << "rootlist:";
  for(int4 i = 0; i < (int4)rootlist.size(); ++i)
    std::cout << ' ' << cg.label_of[rootlist[i]];
  std::cout << std::endl;

  // Walk in the (reordered) component list order so both sides print in the
  // same sequence; labels keep the observation block-name keyed.
  for(int4 i = 0; i < cg.graph.getSize(); ++i) {
    FlowBlock *bl = cg.graph.getBlock(i);
    FlowBlock *dom = bl->getImmedDom();
    std::cout << "idom " << cg.label_of[bl] << ": "
              << (dom == (FlowBlock *)0 ? std::string("NULL") : cg.label_of[dom])
              << std::endl;
  }
}

} // namespace

int main(void)
{
  {
    CaseGraph cg;
    FlowBlock *e = cg.add("E");
    FlowBlock *a = cg.add("A");
    FlowBlock *b = cg.add("B");
    cg.add("O1");
    cg.add("O2");
    cg.edge(e, a);
    cg.edge(a, b);
    run_case("case A: multi-root trailing orphans", cg);
  }
  {
    CaseGraph cg;
    FlowBlock *e = cg.add("E");
    FlowBlock *a = cg.add("A");
    cg.edge(e, a);
    cg.edge(a, e);
    run_case("case B: single root with back-edge into entry", cg);
  }
  {
    // Edge insertion order fixes the in-edge order of M as [R0, R1], which
    // the "first processed predecessor" scan (block.cc:1995-1999) reads.
    CaseGraph cg;
    FlowBlock *r0 = cg.add("R0");
    FlowBlock *m = cg.add("M");
    FlowBlock *x = cg.add("X");
    FlowBlock *r1 = cg.add("R1");
    cg.edge(r0, m);
    cg.edge(r1, m);
    cg.edge(m, x);
    run_case("case C: cross-root merge (virtual-root index alias)", cg);
  }
  {
    CaseGraph cg;
    FlowBlock *x = cg.add("X");
    FlowBlock *y = cg.add("Y");
    cg.edge(x, y);
    cg.edge(y, x);
    run_case("case D: pure cycle, no root candidate", cg);
  }
  return 0;
}

File: tests/oracle/block_domroot_1204.rs
// HTTPD-ADDDESCEND-THROW-0001 fixture — Rust side.
//
// Mirrors tests/oracle/block_domroot_1204.cc line for line against the
// locked oracle (Ghidra_12.0.4_build e40ed13014025f82488b1f8f7bca566894ac376b).
// Exercises BlockGraph::structure_loops (block.cc:2194 -> findSpanningTree
// block.cc:1009-1136, the public 1:1 port that reorders the component list
// to reverse post order) followed by calc_forward_dominator
// (block.cc:1954-2032), plus the side-effect-free build_dom_tree entry
// (local-RPO path used by the non-reordering callers) asserted to produce
// the identical immediate-dominator set.
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

use rugra::address::Address;
use rugra::block::{BlockBasic, BlockGraph, FlowBlock};

type BlockArc = Arc<RwLock<dyn FlowBlock + Send + Sync>>;

struct CaseGraph {
    graph: BlockGraph,
    blocks: Vec<BlockArc>,
    names: Vec<String>,
}

impl CaseGraph {
    fn add(&mut self, name: &str) -> BlockArc {
        // Mirrors the oracle fixture's graph.newBlock() + internal addBlock:
        // one plain block appended to the component vector.
        let bl: BlockArc = Arc::new(RwLock::new(BlockBasic::new(
            self.blocks.len() as i32,
            Address::new(0x1000 + (self.blocks.len() as u64) * 0x10),
        )));
        self.graph.add_block(bl.clone());
        self.blocks.push(bl.clone());
        self.names.push(name.to_string());
        bl
    }

    fn edge(&mut self, from: &BlockArc, to: &BlockArc) {
        self.graph.add_edge(from.clone(), to.clone());
    }

    fn label_of(&self, bl: &BlockArc) -> String {
        let ptr = Arc::as_ptr(bl) as *const () as usize;
        for (i, b) in self.blocks.iter().enumerate() {
            if Arc::as_ptr(b) as *const () as usize == ptr {
                return self.names[i].clone();
            }
        }
        "?".to_string()
    }
}

fn idom_label(cg: &CaseGraph, bl: &BlockArc) -> String {
    let dom = bl.read().unwrap().get_immed_dom().and_then(|w| w.upgrade());
    match dom {
        Some(d) => cg.label_of(&d),
        None => "NULL".to_string(),
    }
}

fn run_case(title: &str, mut cg: CaseGraph) {
    println!("{}", title);

    // Oracle chain: structureLoops(rootlist) + calcForwardDominator(rootlist).
    let mut rootlist: Vec<BlockArc> = Vec::new();
    cg.graph.structure_loops(&mut rootlist).expect("structure_loops");
    cg.graph
        .calc_forward_dominator(&rootlist)
        .expect("calc_forward_dominator");

    print!("rpo:");
    for i in 0..cg.graph.get_size() {
        let bl = cg.graph.get_block(i).expect("block");
        let idx = bl.read().unwrap().get_index();
        print!(" {}({})", cg.label_of(&bl), idx);
    }
    println!();

    print!("rootlist:");
    for r in &rootlist {
        print!(" {}", cg.label_of(r));
    }
    println!();

    for i in 0..cg.graph.get_size() {
        let bl = cg.graph.get_block(i).expect("block");
        println!("idom {}: {}", cg.label_of(&bl), idom_label(&cg, &bl));
    }
}

/// Build an identical fresh copy of the case (same vector order, same edge
/// insertion order) and run the non-reordering build_dom_tree entry,
/// asserting the immediate-dominator set matches the oracle chain's result
/// (this is the entry heritage/coreaction/blockaction glue callers use).
fn assert_build_dom_tree_matches(
    title: &str,
    edges: &[(usize, usize)],
    nblk: usize,
    expected: &[(usize, &str)], // (block ordinal, expected idom label or "NULL")
    name_of: &dyn Fn(usize) -> String,
) {
    let mut cg = CaseGraph {
        graph: BlockGraph::new(),
        blocks: Vec::new(),
        names: (0..nblk).map(|i| format!("B{}", i)).collect(),
    };
    for i in 0..nblk {
        let name = name_of(i);
        cg.names[i] = name;
        let bl: BlockArc = Arc::new(RwLock::new(BlockBasic::new(
            i as i32,
            Address::new(0x1000 + (i as u64) * 0x10),
        )));
        cg.graph.add_block(bl.clone());
        cg.blocks.push(bl);
    }
    for (f, t) in edges {
        let from = cg.blocks[*f].clone();
        let to = cg.blocks[*t].clone();
        cg.graph.add_edge(from, to);
    }
    cg.graph.build_dom_tree();

    let label_by_ptr: HashMap<usize, String> = cg
        .blocks
        .iter()
        .enumerate()
        .map(|(i, b)| (Arc::as_ptr(b) as *const () as usize, cg.names[i].clone()))
        .collect();
    let mut mismatches = Vec::new();
    for &(ordinal, expect) in expected {
        let bl = &cg.blocks[ordinal];
        let dom = bl.read().unwrap().get_immed_dom().and_then(|w| w.upgrade());
        let got = match dom {
            Some(d) => label_by_ptr
                .get(&(Arc::as_ptr(&d) as *const () as usize))
                .cloned()
                .unwrap_or_else(|| "?".to_string()),
            None => "NULL".to_string(),
        };
        if got != expect {
            mismatches.push(format!(
                "{}: idom({}) = {} expected {}",
                title, cg.names[ordinal], got, expect
            ));
        }
    }
    if !mismatches.is_empty() {
        panic!("build_dom_tree divergence: {:?}", mismatches);
    }
    // Silent on success: stdout stays byte-identical to the oracle fixture.
    // The parity coverage itself is registered in the metadata.
    let _ = title;
}

fn main() {
    {
        let mut cg = CaseGraph {
            graph: BlockGraph::new(),
            blocks: Vec::new(),
            names: Vec::new(),
        };
        let e = cg.add("E");
        let a = cg.add("A");
        let b = cg.add("B");
        cg.add("O1");
        cg.add("O2");
        cg.edge(&e, &a);
        cg.edge(&a, &b);
        run_case("case A: multi-root trailing orphans", cg);
    }
    {
        let mut cg = CaseGraph {
            graph: BlockGraph::new(),
            blocks: Vec::new(),
            names: Vec::new(),
        };
        let e = cg.add("E");
        let a = cg.add("A");
        let e2 = e.clone();
        cg.edge(&e, &a);
        cg.edge(&a, &e2);
        run_case("case B: single root with back-edge into entry", cg);
    }
    {
        // Edge insertion order fixes the in-edge order of M as [R0, R1],
        // which the "first processed predecessor" scan (block.cc:1995-1999)
        // reads.
        let mut cg = CaseGraph {
            graph: BlockGraph::new(),
            blocks: Vec::new(),
            names: Vec::new(),
        };
        let r0 = cg.add("R0");
        let m = cg.add("M");
        let x = cg.add("X");
        let r1 = cg.add("R1");
        cg.edge(&r0, &m);
        cg.edge(&r1, &m);
        cg.edge(&m, &x);
        run_case("case C: cross-root merge (virtual-root index alias)", cg);
    }
    {
        let mut cg = CaseGraph {
            graph: BlockGraph::new(),
            blocks: Vec::new(),
            names: Vec::new(),
        };
        let x = cg.add("X");
        let y = cg.add("Y");
        let x2 = x.clone();
        cg.edge(&x, &y);
        cg.edge(&y, &x2);
        run_case("case D: pure cycle, no root candidate", cg);
    }

    // Parity of the non-reordering entry against the same expectations.
    assert_build_dom_tree_matches(
        "case A",
        &[(0, 1), (1, 2)],
        5,
        &[(0, "NULL"), (1, "E"), (2, "A"), (3, "NULL"), (4, "NULL")],
        &|i| ["E", "A", "B", "O1", "O2"][i].to_string(),
    );
    assert_build_dom_tree_matches(
        "case B",
        &[(0, 1), (1, 0)],
        2,
        &[(0, "NULL"), (1, "E")],
        &|i| ["E", "A"][i].to_string(),
    );
    assert_build_dom_tree_matches(
        "case C",
        &[(0, 1), (3, 1), (1, 2)],
        4,
        &[(0, "NULL"), (1, "R0"), (2, "M"), (3, "NULL")],
        &|i| ["R0", "M", "X", "R1"][i].to_string(),
    );
    assert_build_dom_tree_matches(
        "case D",
        &[(0, 1), (1, 0)],
        2,
        &[(0, "NULL"), (1, "X")],
        &|i| ["X", "Y"][i].to_string(),
    );
}
